File: apps/mobile/ios/Uniswap/Onboarding/Import/SeedPhraseInputViewModel.swift
//
//  SeedPhraseInputViewModel.swift
//  Uniswap
//
//  Created by Gary Ye on 9/10/23.
//

import Foundation

class SeedPhraseInputViewModel: ObservableObject {

  enum Status: String {
    case none
    case valid
    case error
  }

  enum MnemonicError {
    case invalidPhrase
    case invalidWord(String)
    case notEnoughWords
    case tooManyWords
    case wrongRecoveryPhrase
    case wordIsAddress
  }

  struct ReactNativeStrings {
    var inputPlaceholder: String
    var pasteButton: String
    var errorInvalidWord: String
    var errorPhraseLength: String
    var errorWrongPhrase: String
    var errorInvalidPhrase: String
    var errorWordIsAddress: String
  }

  let rnEthersRS = RNEthersRS()

  // Following block of variables will come from RN
  @Published var targetMnemonicId: String? = nil

  @Published var testID: String? = nil

  @Published var rawRNStrings: Dictionary<String, String> = Dictionary<String, String>() {
    didSet {
      strings = ReactNativeStrings(
        inputPlaceholder: rawRNStrings["inputPlaceholder"] ?? "",
        pasteButton: rawRNStrings["pasteButton"] ?? "",
        errorInvalidWord: rawRNStrings["errorInvalidWord"] ?? "",
        errorPhraseLength: rawRNStrings["errorPhraseLength"] ?? "",
        errorWrongPhrase: rawRNStrings["errorWrongPhrase"] ?? "",
        errorInvalidPhrase: rawRNStrings["errorInvalidPhrase"] ?? "",
        errorWordIsAddress: rawRNStrings["errorWordIsAddress"] ?? ""
      )
    }
  }
  @Published var strings: ReactNativeStrings = ReactNativeStrings(
    inputPlaceholder: "",
    pasteButton: "",
    errorInvalidWord: "",
    errorPhraseLength: "",
    errorWrongPhrase: "",
    errorInvalidPhrase: "",
    errorWordIsAddress: ""
  )
  @Published var onInputValidated: RCTDirectEventBlock = { _ in }
  @Published var onMnemonicStored: RCTDirectEventBlock = { _ in }
  @Published var onPasteStart: RCTDirectEventBlock = { _ in }
  @Published var onPasteEnd: RCTDirectEventBlock = { _ in }
  @Published var onSubmitError: RCTDirectEventBlock = { _ in }
  @Published var onHeightMeasured: RCTDirectEventBlock = { _ in }

  private var lastWordValidationTimer: Timer?
  private let lastWordValidationTimeout: TimeInterval = 1.0

  @Published var input = "" {
    didSet {
      handleInputChange()
    }
  }

  @Published var isFocused = false

  @Published var skipLastWord = true
  @Published var status: Status = .none
  @Published var error: MnemonicError? = nil

  private let minCount = 12
  private let maxCount = 24

  func handleSubmit() {
    lastWordValidationTimer?.invalidate()
    
    let normalized = normalizeInput(value: input)
    let mnemonic = trimInput(value: normalized)
    let words = mnemonic.components(separatedBy: " ")
    let valid = rnEthersRS.validateMnemonic(mnemonic: mnemonic)
      
    error = nil
    if (words.count < minCount || minCount + 1 ..< maxCount ~= words.count) {
      status = .error
      error = .notEnoughWords
    } else if (words.count > maxCount) {
      status = .error
      error = .tooManyWords
    } else if (!valid) {
      status = .error
      error = .invalidPhrase
    } else {
      submitMnemonic(mnemonic: mnemonic)
    }
    
    if (error != nil) {
      onSubmitError([:])
    }
  }

  private func submitMnemonic(mnemonic: String) {
    if (targetMnemonicId != nil) {
      rnEthersRS.generateAddressForMnemonic(
        mnemonic: mnemonic,
        derivationIndex: 0,
        resolve: { mnemonicId in
          if (targetMnemonicId == String(describing: mnemonicId ?? "")) {
            storeMnemonic(mnemonic: mnemonic)
          } else {
            status = .error
            error = .wrongRecoveryPhrase
            onSubmitError([:])
          }
        }, reject: { code, message, error in
          onSubmitError([:])
          print("SeedPhraseInputView model error while generating address: \(message ?? "")")
        })
    } else {
      storeMnemonic(mnemonic: mnemonic)
    }
  }

  private func storeMnemonic(mnemonic: String) {
    rnEthersRS.importMnemonic(
      mnemonic: mnemonic,
      resolve: { mnemonicId in
        onMnemonicStored(["mnemonicId": String(describing: mnemonicId ?? "")])
      },
      reject: { code, message, error in
        onSubmitError([:])
        print("SeedPhraseInputView model error while storing mnemonic: \(message ?? "")")
      }
    )
  }

  private func normalizeInput(value: String) -> String {
    return value.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression).lowercased()
  }
  
  private func isAddress(value: String) -> Bool {
    return value.starts(with: "0x") && value.count == 42
  }

  private func trimInput(value: String) -> String {
    return value.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private func handleInputChange() {
    let normalized = normalizeInput(value: input)
    let skipLastWord = normalized.last != " "
    let skipInvalidWord = skipLastWord && !isAddress(value: normalized)
    validateInput(normalizedInput: normalized, skipInvalidWord: skipInvalidWord)

    lastWordValidationTimer?.invalidate()

    if (skipLastWord) {
      lastWordValidationTimer = Timer.scheduledTimer(
        withTimeInterval: lastWordValidationTimeout,
        repeats: false) { _ in
            DispatchQueue.global(qos: .background).async {
                self.validateInput(normalizedInput: normalized, skipInvalidWord: false)
            }
        }
    }
  }

  private func validateInput(normalizedInput: String, skipInvalidWord: Bool) {
    let mnemonic = trimInput(value: normalizedInput)
    
    let words = mnemonic.components(separatedBy: " ")

    let isValidLength = words.count >= minCount && words.count <= maxCount
    let firstInvalidWord = rnEthersRS.findInvalidWord(mnemonic: mnemonic)
    
    let isAddress = mnemonic.starts(with: "0x") && mnemonic.count == 42

    if (firstInvalidWord == words.last && skipInvalidWord) {
      status = .none
    } else if (firstInvalidWord == "" && isValidLength) {
      status = .valid
    } else if (isAddress) {
      status = .error
      error = .wordIsAddress
    } else if (firstInvalidWord != "") {
      status = .error
      error = .invalidWord(firstInvalidWord)
    } else {
      status = .none
    }

    if (status != .error) {
      error = nil
    }

    let canSubmit = error == nil && mnemonic != "" && firstInvalidWord == "" && isValidLength
    onInputValidated(["canSubmit": canSubmit])
  }
}

File: apps/mobile/android/app/src/main/java/com/uniswap/onboarding/import/SeedPhraseInputViewModel.kt
package com.uniswap.onboarding.import

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.text.input.TextFieldValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.uniswap.EthersRs
import com.uniswap.RnEthersRs
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SeedPhraseInputViewModel(
  private val ethersRs: RnEthersRs,
  private val onInputValidated: (canSubmit: Boolean) -> Unit,
  private val onMnemonicStored: (mnemonicId: String) -> Unit,
  private val onSubmitError: () -> Unit,
) : ViewModel() {

  sealed interface Status {
    object None : Status
    object Valid : Status
    class Error(val error: MnemonicError) : Status
  }

  sealed interface MnemonicError {
    class InvalidWord(val word: String) : MnemonicError
    object TooManyWords : MnemonicError
    object NotEnoughWords : MnemonicError
    object WrongRecoveryPhrase : MnemonicError
    object InvalidPhrase : MnemonicError
    object WordIsAddress : MnemonicError
  }

  data class ReactNativeStrings(
    val inputPlaceholder: String,
    val pasteButton: String,
    val errorInvalidWord: String,
    val errorPhraseLength: String,
    val errorWrongPhrase: String,
    val errorInvalidPhrase: String,
    val errorWordIsAddress: String,
  )

  // Sourced externally from RN
  var targetMnemonicId by mutableStateOf<String?>(null)
  var rnStrings by mutableStateOf(
    ReactNativeStrings(
      inputPlaceholder = "",
      pasteButton = "",
      errorInvalidWord = "",
      errorPhraseLength = "",
      errorWrongPhrase = "",
      errorInvalidPhrase = "",
      errorWordIsAddress = "",
    )
  )

  var input by mutableStateOf(TextFieldValue(""))
    private set
  var status by mutableStateOf<Status>(Status.None)
    private set
  private var validateLastWordTimeout: Long = 1000
  private var validateLastWordJob: Job? = null

  var isFocused by mutableStateOf(false)
    private set

  fun focus() {
    isFocused = true
  }

  fun blur() {
    isFocused = false
  }

  fun handleInputChange(value: TextFieldValue) {
    input = value
    val normalized = normalizeInput(value)

    val skipLastWord = normalized.lastOrNull() != ' '
    val skipInvalidWord = skipLastWord && !isAddress(normalized)
    validateInput(normalized, skipInvalidWord)

    validateLastWordJob?.cancel()

    validateLastWordJob = viewModelScope.launch(Dispatchers.Default) {
      delay(validateLastWordTimeout)
      validateInput(normalized, false)
    }
  }

  private fun validateInput(normalizedInput: String, skipInvalidWord: Boolean) {
    val mnemonic = normalizedInput.trim()
    val words = mnemonic.split(" ")

    if (words.isEmpty()) {
      status = Status.None
      return
    }

    val isValidLength = words.size in MIN_LENGTH..MAX_LENGTH
    val firstInvalidWord = EthersRs.findInvalidWord(mnemonic)

    status = if (firstInvalidWord == words.last() && skipInvalidWord) {
      Status.None
    } else if (firstInvalidWord.isEmpty() && isValidLength) {
      Status.Valid
    } else if (isAddress(mnemonic)) {
      Status.Error(MnemonicError.WordIsAddress)
    } else if (firstInvalidWord.isNotEmpty()) {
      Status.Error(MnemonicError.InvalidWord(firstInvalidWord))
    } else {
      Status.None
    }

    val canSubmit = status !is Status.Error && mnemonic != "" && firstInvalidWord.isEmpty()
    onInputValidated(canSubmit)
  }

  private fun normalizeInput(value: TextFieldValue) =
    value.text.replace("\\s+".toRegex(), " ").lowercase()

  private fun isAddress(value: String) = value.startsWith("0x") && value.length == 42

  fun handleSubmit() {
    validateLastWordJob?.cancel()

    try {
      val normalized = normalizeInput(input)
      val mnemonic = normalized.trim()
      val words = mnemonic.split(" ")
      val valid = EthersRs.validateMnemonic(mnemonic)

      if (words.size < MIN_LENGTH || words.size in MIN_LENGTH + 1..<MAX_LENGTH) {
        status = Status.Error(MnemonicError.NotEnoughWords)
      } else if (words.size > MAX_LENGTH) {
        status = Status.Error(MnemonicError.TooManyWords)
      } else if (!valid) {
        status = Status.Error(MnemonicError.InvalidPhrase)
      } else {
        submitMnemonic(mnemonic)
      }
    } catch (e: Exception) {
      // TODO gary add production logging and update rust code to convert to Java exceptions
      Log.d("SeedPhraseInputViewModel", "Storing mnemonic caused error ${e.message}")
    }

    if (status is Status.Error) {
      onInputValidated(false)
      onSubmitError()
    }
  }

  private fun submitMnemonic(mnemonic: String) {
    if (targetMnemonicId != null) {
      val generatedId = ethersRs.generateAddressForMnemonic(mnemonic, derivationIndex = 0)
      if (generatedId != targetMnemonicId) {
        status = Status.Error(MnemonicError.WrongRecoveryPhrase)
        onSubmitError()
      } else {
        storeMnemonic(mnemonic)
      }
    } else {
      storeMnemonic(mnemonic)
    }
  }

  private fun storeMnemonic(mnemonic: String) {
    val mnemonicId = ethersRs.importMnemonic(mnemonic)
    onMnemonicStored(mnemonicId)
  }

  companion object {
    private const val MIN_LENGTH = 12
    private const val MAX_LENGTH = 24
  }

}
